cli/db_export.go

```diff
@@ -0,0 +1,158 @@
+// Copyright 2023 Democratized Data Foundation
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package cli
+
+import (
+	"encoding/json"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+
+	"github.com/fxamacker/cbor/v2"
+	"github.com/spf13/cobra"
+
+	httpapi "github.com/sourcenetwork/defradb/api/http"
+	"github.com/sourcenetwork/defradb/config"
+	"github.com/sourcenetwork/defradb/logging"
+)
+
+const cborFileType = "cbor"
+const jsonFileType = "json"
+
+func MakeDBExportCommand(cfg *config.Config) *cobra.Command {
```

Comment:
todo: This looks awfully expensive in terms of memory, With the results being fully held in memory from the http call, unmarshalled, then remarshalled, before being written to file. For all collections at the same time in the database.

Worse, the entire database is copied into memory server-side too, and could do quite a lot of damage.

Whilst we might be okay tolerating this for a short time, I think the command documentation should come with a very large warning. And another should be added to the func documentation of the http export command.

Reply:
I was going to bring this up during standup. There is a simple solution to this which is to let the defra instance write/read the file directly like postgres and other DBs do.

Reply:
Just noting that the CLI or DB behaviour doesnt actually need to change to not load everything into memory. Network stuff might need to change though if you want the total IM size to be less than 1 DB.

Reply:
We could stream every document across one at a time.